Automattic/woocommerce-payments
# UPE basic automatic theming

Fixes #1827 

#### Changes proposed in this Pull Request

Styles the following elements in UPE:

* Inputs: Similar styles as the inputs in WooCommerce form.
* Labels: Similar styles as the labels in WooCommerce form
* Selected Tab: Similar styles as the 'Place order' button.
* Unselected Tabs: Similar styles as the inputs in WooCommerce form.
* Icons: On selected tab, same color as the text in Place order button. On unselected tab, same color as the text on WooCommerce inputs.
* Fonts: Parses the fonts on the page, if it finds fonts from the list below, they are included with UPE:
  * fonts.googleapis.com
  * fonts.gstatic.com
  * fast.fonts.com
  * use.typekit.net

The following themes have been used for testing:
1) Storefront
2) 2021
3) 2020
4) Astra
5) Deli
6) Leto

This function is used to compute the styles: https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle


Known issues:
- When there are saved payment methods, the icons and the dropdown arrow are pushed down. This is coming from the way UPE generates the top property. Stripe is aware and fixing it.
- Not all the icons on the tabs are fillable. Stripe is aware and will change the icons to fillable icons.
- This PR doesn't cover Hover, Focus and Invalid states.
- Given the diversity of existing themes, this is a first iteration. Once merchants start using it, we will probably need to adapt it to work with other themes.








#### Testing instructions





* Enter your WC Options table and add the feature flag _wcpay_feature_upe with a value of 1. Stripe UPE payment gateway should now replace the Credit Card and be enabled at checkout. 
* Please go to Payments > Settings and ensure that WooCommerce Payments is enabled.
* Add an item from your shop to your cart and proceed through a regular checkout. At checkout, select "WooCommerce Payments" and you should see the UPE appear below the payment gateway. 
* Currently there is only one payment method active, cards, so not tabs are selected. To see the tabs, add the payments to [this array](https://github.com/Automattic/woocommerce-payments/blob/634d88056b48fd18f81787ea03bb597f24f3987f/includes/payment-methods/class-upe-payment-gateway.php#L110): [ 'bancontact', 'card', 'ideal' ]

Check that UPE tabs, inputs and labels follow the rules described above.
A few screenshots:
<img width="421" alt="UPEStorefront" src="https://user-images.githubusercontent.com/8002881/119880698-1fcfb080-bef2-11eb-92e2-96e5af60a650.png">
<img width="795" alt="UPE2021" src="https://user-images.githubusercontent.com/8002881/119880704-21997400-bef2-11eb-9b05-d0fc4978f437.png">
<img width="440" alt="UPEDeli" src="https://user-images.githubusercontent.com/8002881/119880714-23fbce00-bef2-11eb-924b-a03c9471daef.png">


-------------------

- [x] Tested on mobile (or does not apply) -> Tested but not comprehensively.
